src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Stream.cs

```diff
@@ -406,36 +404,38 @@ public static partial class JsonSerializer
                 ThrowHelper.ThrowArgumentNullException(nameof(jsonTypeInfo));
             }
 
-            return CreateAsyncEnumerableDeserializer<TValue>(utf8Json, jsonTypeInfo, cancellationToken);
+            return CreateAsyncEnumerableDeserializer(utf8Json, CreateQueueTypeInfo<TValue>(jsonTypeInfo), cancellationToken);
+        }
+
+        private static JsonTypeInfo<Queue<TValue>> CreateQueueTypeInfo<TValue>(JsonTypeInfo jsonTypeInfo)
+        {
+            return JsonMetadataServices.CreateQueueInfo<Queue<TValue>, TValue>(
+                options: jsonTypeInfo.Options,
+                collectionInfo: new()
+                {
+                    ObjectCreator = static () => new Queue<TValue>(),
+                    ElementInfo = jsonTypeInfo,
+                    NumberHandling = jsonTypeInfo.Options.NumberHandling
+                });
         }
 
         private static async IAsyncEnumerable<TValue> CreateAsyncEnumerableDeserializer<TValue>(
             Stream utf8Json,
-            JsonTypeInfo jsonTypeInfo,
+            JsonTypeInfo<Queue<TValue>> queueTypeInfo,
             [EnumeratorCancellation] CancellationToken cancellationToken)
         {
-            JsonSerializerOptions options = jsonTypeInfo.Options;
-            JsonTypeInfo<Queue<TValue>> queueTypeInfo =
-                JsonMetadataServices.CreateQueueInfo<Queue<TValue>, TValue>(
-                    options: options,
-                    collectionInfo: new()
-                    {
-                        ObjectCreator = () => new Queue<TValue>(),
-                        ElementInfo = jsonTypeInfo,
-                        NumberHandling = options.NumberHandling
-                    });
-
+            queueTypeInfo.EnsureConfigured();
+            JsonSerializerOptions options = queueTypeInfo.Options;
             var bufferState = new ReadBufferState(options.DefaultBufferSize);
             ReadStack readStack = default;
-            queueTypeInfo.EnsureConfigured();
             readStack.Initialize(queueTypeInfo, supportContinuation: true);
             var jsonReaderState = new JsonReaderState(options.GetReaderOptions());
 
             try
             {
                 do
                 {
-                    bufferState = await ReadFromStreamAsync(utf8Json, bufferState, cancellationToken).ConfigureAwait(false);
+                    bufferState = await bufferState.ReadFromStreamAsync(utf8Json, cancellationToken, fillBuffer: false).ConfigureAwait(false);
```

Comment:
Can this use the new Stream.ReadAtLeastAsync method when targeting .NET 7?

Reply:
> Can this use the new Stream.ReadAtLeastAsync method when targeting .NET 7 and fillBuffer is true?

It could, but we still need to polyfill for all the other TFMs. Does it make a difference perf-wise?

> And if fillBuffer is false, why even call this method rather than just calling ReadAsync directly?

The method does a bit of added bookkeeping like skipping BOM bytes and tracking the maximum watermark in the rented buffers.

Reply:
> Does it make a difference perf-wise?

No. It'd just be nice to use the built-in functionality where it exists. We could implement the same method as an extension polyfill for downlevel.

Reply:
> The method does a bit of added bookkeeping like skipping BOM bytes and tracking the maximum watermark in the rented buffers.

But it's also more expensive: if the stream.ReadAsync call completes asynchronously, it'll force this extra level of async method to suspend and allocate.

Reply:
> > Does it make a difference perf-wise?
> 
> No. It'd just be nice to use the built-in functionality where it exists. We could implement the same method as an extension polyfill for downlevel.

I gave it a shot, but honestly I don't feel like it's improving the codebase. Netstandard2.0 doesn't have `Stream.ReadAsync` overloads that accept `Memory<T>` so you can't really polyfill the exact `ReadAtLeastAsync` method signature (`ArraySegment<byte>` seeming like the next best thing), so it all becomes messier than necessary.